Feature/highlighting of viewer objects

weilbith opened the pull request:
Closes #115
Replaces #116

This brings the feature to detect if the user of the plugin (the `viewer` from the naming of the GH API) has authored an object (e.g. issue or comment) or has reacted to such one. In this case the username or the reaction gets highlighted differently. To do so it puts the usernames into a bubble as it does for labels or reactions. To get an idea how this looks like, checkout the example screenshot below.

While the actual feature additions are quite small, the code changes have become a bit bigger. One part of it comes from the fact that the actual bubbly highlighting implementation was not fully correct working. The delimiter had a fixed color. I always thought something is broken on my setup, but it was the implementation. Also it was duplicated code all other the place. I have iterated over my code over and over again to generalizing code, making it reusable and add small features here and there around bubbles. After all think the feature of drawing bubbles in the UI is now a pretty neat feature that can be used everywhere with ease (as I kinda have proven :grimacing: )

Generally I recommend to review my PRs commit by commit. I try to bundle them always meaningful. Also I take commit messages very serious and explain why a change was needed (for) and what the approach looks like. Everything that even good readable can't (or shouldn't) express. This should enable the reviewer to focus on the code itself, rather then guessing what the higher level intention/reasoning was. Also it is very helpful in all-day development in your editor to read up commit messages for code under the cursor. I hope all decision are clear and you agree with them. Else each commit should be so focused that we can easily delete single history entries or change them with `--fixup` commits.

**Attention:** While I tried to optimize the "bubbeling" as my as I can, I had to drop the bubbles in the quickfix window. Because of bug in the old implementation, it was basically not possibly to keep them like they were. More details as said in the commit messages.

![octo_demo](https://user-images.githubusercontent.com/12543647/109982012-c6485300-7d01-11eb-8ac3-5dc81a26e3e6.png)
(in the screenshot at the bottom there is an example of reactions and their highlighting if the user took part in one of them. Unfortunately they smiley does not work in my terminal. But just that you know.

pwntester commented on lua/octo/writers.lua:
```diff
@@ -1,6 +1,7 @@
 local hl = require "octo.highlights"
```
I think this is no longer used and can be removed

weilbith replied:
Ups. I really need to get my Lua language server running to get such diagnostics. Stupid lazy developer. :laughing:

weilbith replied:
:heavy_check_mark:

pwntester commented on lua/octo/init.lua:
```diff
@@ -149,6 +149,7 @@ function M.create_buffer(type, obj, repo, create)
   end
 
   api.nvim_set_current_buf(bufnr)
+  api.nvim_set_hl_ns(constants.OCTO_HIGHLIGHT_NS)
```
The highlights in the issue title (line 1) dissapeared. May it be related with this?

weilbith replied:
This is required because I set the generic hightlight groups for the bubbles with the Lua API. Such highlight groups take only effect in their namespace. As you had this namespace already declared, I was just using them. 
I could not set the highlight without the API, as I get the color of the bubbles body highlight group via the API too. And these values are confusing. Maybe they are HEX into RGB (as the function parameter suggests) and then all decimal numbers added or concatenated, but it is of type number, ... The docs don't say much about it. Passing them back into the API to set a color works (as I do).
I failed to get the color of a highlight group with Vim script commands. At least I fail to save the response of a VimL command into a lua variable. `local foo = vim.cmd("echo 'bar'")` does not work...
This is also the reason I do not use your `create_highlight` function, as of the incompatible color value schemes. I added a comment there to investigate it one day and fix. But maybe you can help?

pwntester replied:
> This is required because I set the generic hightlight groups for the bubbles with the Lua API. Such highlight groups take only effect in their namespace. As you had this namespace already declared, I was just using them. 

Its ok, I was just trying to understand how `nvim_set_hl` works. I was and still not 100% sure if these HL groups are activated at the buffer, window or vim level. But it seems they are applied at the vim level? also not sure what happens if the activated NS does not define some HL groups, may that be the reason why Title virtual text is no longer highlighted?

> I could not set the highlight without the API, as I get the color of the bubbles body highlight group via the API too. And these values are confusing. Maybe they are HEX into RGB (as the function parameter suggests) and then all decimal numbers added or concatenated, but it is of type number, ... The docs don't say much about it. Passing them back into the API to set a color works (as I do).

They appear to be the int version of the RGB hex codes. So for example 255 is blue (0x0000FF)

> I failed to get the color of a highlight group with Vim script commands. At least I fail to save the response of a VimL command into a lua variable. `local foo = vim.cmd("echo 'bar'")` does not work...

What function were you trying to call? You should be able to do `local foo = vim.fn.some_function()`

> This is also the reason I do not use your `create_highlight` function, as of the incompatible color value schemes. I added a comment there to investigate it one day and fix. But maybe you can help?

Im ok using the new `nvim_set_hl` API, but lets figure out why we lost the title virtual text highlights

pwntester replied:
> Im ok using the new nvim_set_hl API, but lets figure out why we lost the title virtual text highlights

I still dont know what part of this PR causes it, but the reason that Im losing the highlights in the Title line is that sign linehl has now more priority than the other hl groups. So we are also losing markdown highlighting because it gets clobbered with the sign linehl (`OctoNvimEditable`)

pwntester replied:
> I solved this issue at the moment with using a highlight for `OctoNvimEditable` that has only a background

Worked for me! Perhaps we should add a note in the README for this HL group
Other than that, this PR looks great! thank you so much! I think we can rebase and merge :tada:

pwntester commented:
> Generally I recommend to review my PRs commit by commit. I try to bundle them always meaningful. Also I take commit messages very serious and explain why a change was needed (for) and what the approach looks like. Everything that even good readable can't (or shouldn't) express. This should enable the reviewer to focus on the code itself, rather then guessing what the higher level intention/reasoning was. Also it is very helpful in all-day development in your editor to read up commit messages for code under the cursor. I hope all decision are clear and you agree with them. Else each commit should be so focused that we can easily delete single history entries or change them with `--fixup` commits.

It looks great, will try to do the same from now on

> **Attention:** While I tried to optimize the "bubbeling" as my as I can, I had to drop the bubbles in the quickfix window. Because of bug in the old implementation, it was basically not possibly to keep them like they were. More details as said in the commit messages.

Thats ok for now, we will see if we can add them later or even create our own QF-like buffer

pwntester commented:
Sometimes it feels like there are too many bubbles in my opinion. I like that I can tweak the hl groups to make the bubble color the same as the Normal background but now its all or nothing for Viewer bubbles. Can you make a hl group for the viewer reactions, and a different one for viewer name? or is it too much HL groups :thinking_face:

Also:
- I love the viewer icon, but we can make it configurable
- what do you think about adding the viewer icon to viewer reaction groups?)

weilbith commented on README.md:
```diff
@@ -175,37 +175,42 @@ Just edit the issue title, description or comments as a regular buffer and use `
+| `OctoNvimPullModifications`  | GitHub color  |
+
+The term `GitHub color` refers to the colors used in the WebUI.
+The `viewer` means the user of the plugin or more precisely the user authenticated
+via the `gh` CLI tool used to retrieve the data from GitHub.
```
Oh sorry, I tend to format Markdown with `gq` to a textwidth of 78 characters. But GH respects such newlines right? I'll fix that.

weilbith replied:
:heavy_check_mark:

weilbith commented on lua/octo/writers.lua:
```diff
@@ -1,6 +1,7 @@
 local hl = require "octo.highlights"
 local constants = require "octo.constants"
 local util = require "octo.util"
+local bubbles = require "octo.ui.bubbles"
```
Yes. This is the general intention. But I don't want to do too much, especially not in this PR. It is rather just the "kickoff" for some change that takes place over time. I like idea to always leave the code better than it was before without make super broad changes at once.
Therefore I'll close this conversation. Please re-open when you strongly disagree.

weilbith commented on lua/octo/writers.lua:
```diff
@@ -574,15 +584,23 @@ function M.write_review_thread_header(bufnr, opts, line)
-    table.insert(header_vt, {" ", "OctoNvimBubbleDelimiter"})
+    local outdated_bubble = bubbles.make_bubble(
+      "outdated",
+      "OctoNvimBubbleGreen",
```
Oh damn. That slipped through. At some point of refactoring the bubble commands I must made a stupid copy-paste. Sorry for that. I'll fix that.

weilbith replied:
:heavy_check_mark:

weilbith commented on lua/octo/writers.lua:
```diff
@@ -148,10 +152,13 @@ function M.write_details(bufnr, issue, update)
-    {issue.author.login, "OctoNvimDetailsValue"}
-  }
+  local author_vt = {{"Created by: ", "OctoNvimDetailsLabel"}}
+  local author_bubble = bubbles.make_user_bubble(
```
You can decide. I mean I added them because I think it is nice. That's why it is there.
In first place it was not the intention to just bubble it, but highlight the author if it is the viewer himself. And as the colored bubble was the choice of doing so... Though I still like the general bubble. :grimacing: 
So if you don't like it, I can just remove that. Else I'll add them for the other locations as well. Please note your decision. :upside_down_face:

pwntester replied:
Lets keep it but then, lets add it to assignees and reviewers as well for consistency and to highlight what is the viewer role in the issue/pr

weilbith replied:
Okay, I extended it for all. And with the additional style configuration options you can "disable" it where ever you like.
One note: PR reviewers do not seem to work atm. The API always states there are no.

weilbith commented:
> Sometimes it feels like there are too many bubbles in my opinion. I like that I can tweak the hl groups to make the bubble color the same as the Normal background but now its all or nothing for Viewer bubbles. Can you make a hl group for the viewer reactions, and a different one for viewer name? or is it too much HL groups :thinking_face:

Haha. I guess that is (Neo)Vim. Everyone has his own taste and expresses it. I like it.
Nice workaround you found there. Should we document it as hint for users who feel the same.
Yes I think we could add such a highlight group. Maybe we could have the low level classes as now and then add more specific ones. These will just link to their "parents", but are open for local adoption for each user who wants to tweak it that way. I don't mind to have so many highlight groups. Setting them as user is 100% optional, we have sane defaults, and having such a highlight group hierarchy  allows to set single groups with a wide effect without loosing the ability to absolutely tweak it.
 
> Also:
> 
>     * I love the viewer icon, but we can make it configurable

On my list of TODOs there is already the task to make all icons configurable. Also reactions (e.g. yours do not work for me). I'm thinking of a more generic interface that makes adding new icons/emojis very easy and flexible. But if you like to I could still add a global scope variable just for now? Would this be helpful?

>     * what do you think about adding the viewer icon to viewer reaction groups?)

Aeh? I don't know what you mean. :see_no_evil: 
Could you describe it in more detail (maybe even just send the code references as links) or send a screenshot? :pray:

pwntester commented:
> Maybe we could have the low level classes as now and then add more specific ones. These will just link to their "parents", but are open for local adoption for each user who wants to tweak it that way. I don't mind to have so many highlight groups. Setting them as user is 100% optional, we have sane defaults, and having such a highlight group hierarchy allows to set single groups with a wide effect without loosing the ability to absolutely tweak it.

LGTM 👍🏻 

> (e.g. yours do not work for me)

What terminal emulator do you use?

> But if you like to I could still add a global scope variable just for now?

No, its ok, we can wait

> Could you describe it in more detail (maybe even just send the code references as links) or send a screenshot? 🙏

Something like this (also removed the padding):

<img width="725" alt="image" src="https://user-images.githubusercontent.com/125701/110170788-ffb0b980-7dfa-11eb-94f9-70d216cfa4aa.png">

weilbith commented:
> What terminal emulator do you use?

I use Alacritty with TMux inside with Source Code Pro Nerd Font.

> Something like this (also removed the padding):

Hmm. I added the icon to preceed any username to symbolize: this is a user. So atm this icon is specific to these items, no matter if its the viewer or not. To show the user if it is him or someone else is done by the color only. For reactions, which are no users, there is the icon of the reaction itself and then again the coloring depending of the viewer.
If I interpret your image correctly you now want to add the user icon to a reaction if the reaction was by/includes the viewer plus colering it? Is that right?
If you really like the idea about having a second icon in reactions by the viewer, we might want to add another icon that can be customized and make it a unicode dot? Just like a tiny little hint as you do with notification icons when there are new notifications. And if you like to have it the same as the user icon, go for it.
Sorry, for making these comment so long. Just want to make the async communication as efficient as possible by providing always enough information. :see_no_evil:

---

Could you please provide input to the left open review conversations? :upside_down_face:

weilbith commented:
Off-topic, but could you recommend a plugin (might be terminal related) to show images more efficiently? For your image I just used `gx`. That downloads the image and then uses my default web-browser to display it...

pwntester commented:
> Hmm. I added the icon to preceed any username to symbolize: this is a user. So atm this icon is specific to these items, no matter if its the viewer or not. To show the user if it is him or someone else is done by the color only. For reactions, which are no users, there is the icon of the reaction itself and then again the coloring depending of the viewer.
If I interpret your image correctly you now want to add the user icon to a reaction if the reaction was by/includes the viewer plus colering it? Is that right?

I was suggesting adding the icon instead of the coloring, but forget it, coloring it is :ok:
Just lets create a different sub HL group for it

pwntester commented:
> Off-topic, but could you recommend a plugin (might be terminal related) to show images more efficiently? For your image I just used gx. That downloads the image and then uses my default web-browser to display it. Didn't TMux introduce floating windows as well? I mean with Ranger I can view images in Alacritty already.

I also use `gx`. You can try something like https://github.com/seebye/ueberzug

pwntester commented:
Off-topic: I saw you contributed to Packer.nvim. Do you know if this configuration is correct?:
```lua
  use {'~/Dev/octo.nvim', config = function()
    require('telescope').load_extension('octo')
    vim.g.octo_icon_reaction_viewer_hint = ""
  end}
  use {'~/Dev/octo-notifications.nvim', after='octo.nvim', config = function()
```

For some reason, the reaction viewer icon is not picked up. But it does work if I do `let g:octo_icon_reaction_viewer_hint = ""` when starting vim

pwntester commented:
> For some yet unknown reason the API always states that there are no reviewers for any PR. 

The API is actually returning the __requested reviewers__, that is the users that the PR author has requested a review from.
These are the ones that in the WebUI appear with a yellow dot.

If we want to have something similar to what the WebUI shows, then we need to retrieve `pullRequestReviews.author.login` and `pullRequestReview.state`. Also, since I user can leave multiple reviews, we need to "merge" them, so we only show the "strongest" one (eg: if a user leaves `CHANGES_REQUESTED` and `COMMENTED`, only `CHANGES_REQUESTED`). I think that is what the WebUI does, but thats probably work for a new PR

weilbith commented:
I added the asked for FAQ section. I no gonna rebase the whole stuff to integrate the fixup commits and resolve the merge conflicts. :crossed_fingers:

weilbith commented:
Alright that should be it. If you might have a final look we can finally merge it. Sorry that this tiny feature took so long. :see_no_evil:

weilbith commented on README.md:
```diff
@@ -181,37 +181,66 @@ Just edit the issue title, description or comments as a regular buffer and use `
+highlight group that linkes per default to the main bubble highlight group. To
+disable most bubbles at once you can simply link `OctoNvimBubble` to `Normal`.
+To only disable them for a certain plain do the same for the specific sub-group
+(e.g. `OctoNvimUser`)
```
I had it like this before. But I kinda felt stupid to name everything that shows in a bubble with `OctoNvimBubble*`. Especially because we might use a bubble now, but eventually will change for some of them so a different design approach. It would be stupid to have to change these names, just we were too specific about them in the beginning. I think leaving this "detail" out in the name makes them not less readable/understandable, but would just restrict/hurt us on future changes. Do you agree?

pwntester replied:
Ok, lets leave it AS IS for now. We should be good for merging! :tada:

weilbith commented on README.md:
```diff
@@ -181,37 +181,66 @@ Just edit the issue title, description or comments as a regular buffer and use `
+**Why do my issue titles do not get highlightes properly?**
+
+The title of an issue or PR is special as it gets special highlighting applied
+and is an editable section. Due to the latter property it gets the
```
:heavy_check_mark:

weilbith commented on README.md:
```diff
@@ -181,37 +181,66 @@ Just edit the issue title, description or comments as a regular buffer and use `
+
+**Why do my issue titles do not get highlightes properly?**
+
+The title of an issue or PR is special as it gets special highlighting applied
```
:heavy_check_mark:

weilbith commented on README.md:
```diff
@@ -181,37 +181,66 @@ Just edit the issue title, description or comments as a regular buffer and use `
+To only disable them for a certain plain do the same for the specific sub-group
+(e.g. `OctoNvimUser`)
+
+**Why do my issue titles do not get highlightes properly?**
```
:heavy_check_mark:

weilbith commented:
Quick note: PR review comments that I just accept and will do like they are, usually get a thumbs up emoji by me. If I then have pushed the changes and there was no room for interpretation and it was straight forward to apply the requested changes, I put a :heavy_check_mark:  emoji as comment and close the conversation. Just that you know.

pwntester commented:
Thanks for the awesome PR and for helping to set a better architecture and design!

weilbith commented:
@pwntester sorry, I could have communicated that better, but you just merged it with a fixup commit in it. :see_no_evil: 
You eventually want to resolve this and force push to master? Or just keep it as an accident. :laughing:

pwntester commented:
Had to google for what a `fixup!` commit is 🙈 
How can I fix it?

weilbith commented:
Well, there is the chance that a user already updates to latest master and depending on this plugin manager it will then break. He has to remove the "cache" of the plugin manually and re-install. 
The procedure for this very special case would be:

```sh
git switch master
git rebase --interactive 9397ff6~1
# :wq the vim buffer
git push --force-with-lease
```

Unfortunately this got merged with a mergecommit. Using the HEAD~2 reference will make this a huge rebase we don't want. But my specific commands will remove the merge commit itself. This works as I had rebased my PR on top of master before your last review. So we are kinda lucky.
Decide if you really want to fix this history glitch with all its downsides or accept it and learn for future. :stuck_out_tongue: 
I'm very sorry. :see_no_evil:

pwntester commented:
Thanks for the info, I think I will leave it as is since the git history is already messy and kind of dirty. I hope to improve it in the future though 😄

weilbith commented:
I totally agree. With the part to leave it as it, not saying your history is messy. :rofl:
